### ISSUE TITLE
Gradient zero when diff is clipped

### ISSUE THREAD

**Issue by davidsj** [spragunr/deep_q_rl#46]
[This code](https://github.com/spragunr/deep_q_rl/blob/master/deep_q_rl/q_network.py#L102) in q_network.py,
```python
if self.clip_delta > 0:
    diff = diff.clip(-self.clip_delta, self.clip_delta)
```
results in gradients going to zero whenever the diff is clipped. In other words, gradients vanish precisely when they should be largest.

Simplified example of the problem:
```python
>>> a = T.arange(-2, 2, 0.25)
>>> loss = T.sum(a.clip(-1, 1) ** 2)
>>> theano.gradient.grad(loss, a).eval()
array([-0. , -0. , -0. , -0. , -2. , -1.5, -1. , -0.5,  0. ,  0.5,  1. ,
        1.5,  2. ,  0. ,  0. ,  0. ], dtype=float32)
```

The Nature paper is actually misleading in its discussion of clipping:

> We also found it helpful to clip the error term ... to be between -1 and 1. Because the absolute value loss function |x| has a derivative of -1 for all negative values of x and a derivative of 1 for all positive values of x, clipping the squared error to be between -1 and 1 corresponds to using an absolute value loss function for errors outside of the (-1,1) interval.

First, clipping the "squared error" to (-1, 1) isn't even sensible since the squared error is always positive. They presumably mean clipping the error, but even that is wrong since, as we see above, that would incorrectly lead to a zero gradient.

In their Torch implementation they use their clipped error term *directly as the derivative* of an implicit loss function with respect to network output, and propagate *that* backwards to determine the gradient. So their loss, implicitly, is quadratic when error is in (-1, 1), and linear (rather than flat) outside of that region.

I have a patch for this, which I'm currently testing to verify improved learning performance. Results are promising so far.

**Comment by davidsj** [comment]
PR #47 has the fix. The new behavior in simplified form:
```python
>>> clip_delta = 1.
>>> diff = T.arange(-2, 2, 0.25)
>>> quadratic_part = T.minimum(abs(diff), clip_delta)
>>> linear_part = abs(diff) - quadratic_part
>>> loss = T.sum(0.5 * quadratic_part ** 2 + clip_delta * linear_part)
>>> theano.gradient.grad(loss, diff).eval()
array([-1.  , -1.  , -1.  , -1.  , -1.  , -0.75, -0.5 , -0.25,  0.  ,
        0.25,  0.5 ,  0.75,  1.  ,  1.  ,  1.  ,  1.  ], dtype=float32)
```

Or with clip_delta = 1.5:
```python
>>> clip_delta = 1.5
>>> diff = T.arange(-2, 2, 0.25)
>>> quadratic_part = T.minimum(abs(diff), clip_delta)
>>> linear_part = abs(diff) - quadratic_part
>>> loss = T.sum(0.5 * quadratic_part ** 2 + clip_delta * linear_part)
>>> theano.gradient.grad(loss, diff).eval()
array([-1.5 , -1.5 , -1.5 , -1.25, -1.  , -0.75, -0.5 , -0.25,  0.  ,
        0.25,  0.5 ,  0.75,  1.  ,  1.25,  1.5 ,  1.5 ], dtype=float32)
```

I'll post some results on six games tomorrow morning.

**Comment by spragunr** [comment]
Nice work!  I'll merge PR#47 immediately.  The current behavior is clearly broken.  I'll close this issue when we get some confirmation that this improves performance.

**Comment by davidsj** [comment]
The data I've collected so far are depicted in these graphs, which depict reward_per_episode using both the old and new code, for six games up to about 40 epochs.

Results are clearly much better for asterix, breakout, gravitar, seaquest, and space_invaders. Not noticeably better, and possibly worse, for tennis (but tennis is really bad to begin with, and this is not out of family with the Nature paper).

![asterix](https://cloud.githubusercontent.com/assets/10724404/10123061/ea2bd880-64ea-11e5-955f-5c6d5f6c3a74.png)
![breakout](https://cloud.githubusercontent.com/assets/10724404/10123063/ea2f088e-64ea-11e5-83a3-0a26cef0aa00.png)
![gravitar](https://cloud.githubusercontent.com/assets/10724404/10123064/ea2fda02-64ea-11e5-8a7e-71dbdb023266.png)
![seaquest](https://cloud.githubusercontent.com/assets/10724404/10123062/ea2d80b8-64ea-11e5-8f47-84d9e383d00d.png)
![space_invaders](https://cloud.githubusercontent.com/assets/10724404/10123066/ea317fd8-64ea-11e5-8582-5a3ed245aab4.png)
![tennis](https://cloud.githubusercontent.com/assets/10724404/10123065/ea30da10-64ea-11e5-9bc0-901f5e1e3116.png)

**Comment by spragunr** [comment]
Results look promising.  I don't remember seeing the .5 multiplier in the loss function in either the Nature paper or the DQN code.  Am I missing something?

**Comment by davidsj** [comment]
The trick is that the DeepMnd implementation never actually explicitly defines the loss function, it defines the derivative of loss with respect to Q output, and loss is implicit from that. I included the 1/2 to make our explicit loss equal their implicit loss.

The Nature paper incorrectly describes what they do (explained in this issue description). A faithful implementation of the paper, as opposed to their code, results in this very bug. :)

Anyways, RMSProp is pretty scale-invariant so in practice it shouldn't make much difference.

**Comment by spragunr** [comment]
I'm thinking this is a better match to the loss function in the Nature paper: (?)

    >>> clip_delta = 1.0
    >>> diff = T.arange(-2, 2, 0.25)
    >>> quadratic_part = T.minimum(abs(diff), clip_delta/2.0)
    >>> linear_part = abs(diff) - quadratic_part
    >>> loss = T.sum(quadratic_part ** 2 + clip_delta * linear_part)
    >>> theano.gradient.grad(loss, diff).eval()
        array([-1. , -1. , -1. , -1. , -1. , -1. , -1. , -0.5,  0. ,  0.5,  1. ,
            1. ,  1. ,  1. ,  1. ,  1. ], dtype=float32)

**Comment by spragunr** [comment]
Just saw your message as I posted mine.  I agree it's more important to match the behavior of their working code than the errors in the paper :)   I also agree a difference like this probably won't make much difference to learning outcomes, but I think it is actually important to get these details to match.   As much as possible, people should feel comfortable making apples-to-apples comparisons between the behavior of this code and previously published results.

**Comment by davidsj** [comment]
Well, their code sets the derivative to the clipped delta. If you integrate that, you get the loss I specified.

**Comment by spragunr** [comment]
Got it.  Thanks again for catching this.

**Comment by davidsj** [comment]
For reference, 100 epochs for comparison on those six games.

Note that asterix learning going unstable around epoch 70 is consistent with DeepMind's reported results here: http://arxiv.org/abs/1509.06461

![asterix](https://cloud.githubusercontent.com/assets/10724404/10244880/65f7db92-68b7-11e5-8690-26a55ff933a3.png)
![breakout](https://cloud.githubusercontent.com/assets/10724404/10244885/684fa398-68b7-11e5-94d9-9ce1d34aa503.png)
![gravitar](https://cloud.githubusercontent.com/assets/10724404/10244887/69edd490-68b7-11e5-9e1e-948c0a319656.png)
![seaquest](https://cloud.githubusercontent.com/assets/10724404/10244890/6b8f1d54-68b7-11e5-8eb0-cbb59d76b676.png)
![space_invaders](https://cloud.githubusercontent.com/assets/10724404/10244893/6cebedda-68b7-11e5-89af-8a62d1df5859.png)
![tennis](https://cloud.githubusercontent.com/assets/10724404/10244894/6e263de0-68b7-11e5-9f94-9d08b92bdd0e.png)
